### ISSUE TITLE
[question] Myelin: attention fusion and FlashAttention

### ISSUE THREAD

**Issue by vadimkantorov** [NVIDIA/TensorRT#3243]
Hi! When attention op gets fused in a single op with Myelin, it's not written in trex-tooltip if it's using FlashAttention / proper fusion or not (and if it's using quantization under the hood, especially for the implicit quantization mode). How can we know if it's using fused attention impl like FlashAttention? Thanks :)

**Comment by zerollzeng** [comment]
@nvpohanh ^ ^

**Comment by nvpohanh** [comment]
For now,  you can only check the Nsight Systems profiles: https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#nvprof

If the MHA is fused, there should be kernel names with `_mha` in the profile.

In next TRT version, you will be able to get this info by using the IEngineInspector (or `--profilingVerbosity=detailed --dumpLayerInfo` if you're using trtexec).

**Comment by nvpohanh** [comment]
> (and if it's using quantization under the hood, especially for the implicit quantization mode)

TRT's MHA fusion does not support implicit quantizations yet. Please use explicit quantization instead: Add Q/DQ ops before the two batch gemms in the MHA and also add Q/DQ ops before the ResidualAdd.

Here is an ugly example:
![2023-07-25 10_19_35-Window](https://github.com/NVIDIA/TensorRT/assets/53919306/70a9afaf-69d7-4b14-b8d4-47fe891dbbfa)

**Comment by Jeremalloch** [comment]
Hey, a couple questions to tack on:

1. What version of TensorRT will this be included in? Is that TRT 8.7?  And if `--profilingVerbosity=detailed` dumps the data, does that mean using TRT explorer / TREx tooling will show the kernels (instead of a single myelin kernel)?  I'd like to be able [paritially] quantize a model, and being able to see which kernels are executing in INT8 vs FP16 would be very helpful.
2. Does TRT support int8 flash attention?
3. In your attached figure, should we be inserting a QDQ node between the bias addition and the residual connection addition, or should we just be inserting it on the identity / shortcut branch? For ResNet style CNN architectures, I know its recommended to not insert on the residual side so that a conv -> add pattern can be fused into a single kernel. Is this the case for a linear layer -> add in transformers as well? (TRT [doc link](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/graphics/sub-optimal.PNG))

**Comment by nvpohanh** [comment]
The INT8 MHA fused kernels are already integrated in TRT 8.6. The only caveat is that SeqLen must be 512 or below.

It does use flash attention if applicable.

> In your attached figure, should we be inserting a QDQ node between the bias addition and the residual connection addition, or should we just be inserting it on the identity / shortcut branch? For ResNet style CNN architectures, I know its recommended to not insert on the residual side so that a conv -> add pattern can be fused into a single kernel. Is this the case for a linear layer -> add in transformers as well?

For Transformers, it is recommended to add Q/DQs on both the inputs of the ResidualAdd. This is because in ConvNets, we fuse the ResidualAdd with the Conv, but for Transformers, we fuse the ResidualAdd with the LayerNorm that comes right after the ResidualAdd.

**Comment by Jeremalloch** [comment]
Hey, one followup question. How does the fusion scheme work for pre-norm transformers (as the layer norm would only be applied to the residual branch, and not the identity branch)?  Does a norm first transformer come with a performance penalty / less optimized kernels?
![Screenshot 2023-09-21 at 3 49 46 PM](https://github.com/NVIDIA/TensorRT/assets/16086901/8e16d773-cdb2-4661-b719-4f15dff339d1)

**Comment by nvpohanh** [comment]
TRT should be able to fuse the add_1 with norm2, so it should not cause any perf issue.

**Comment by WeixiangXu** [comment]
> TRT's MHA fusion does not support implicit quantizations yet. Please use explicit quantization instead: Add Q/DQ ops before the two batch gemms in the MHA and also add Q/DQ ops before the ResidualAdd.

@nvpohanh How about the int8 attention speed v.s. fp16? Do you have more detailed documents about how to conduct explicit quantization on attention? thanks!

**Comment by zhexinli** [comment]
> The INT8 MHA fused kernels are already integrated in TRT 8.6. The only caveat is that SeqLen must be 512 or below.
> 
> It does use flash attention if applicable.
> 
> > In your attached figure, should we be inserting a QDQ node between the bias addition and the residual connection addition, or should we just be inserting it on the identity / shortcut branch? For ResNet style CNN architectures, I know its recommended to not insert on the residual side so that a conv -> add pattern can be fused into a single kernel. Is this the case for a linear layer -> add in transformers as well?
> 
> For Transformers, it is recommended to add Q/DQs on both the inputs of the ResidualAdd. This is because in ConvNets, we fuse the ResidualAdd with the Conv, but for Transformers, we fuse the ResidualAdd with the LayerNorm that comes right after the ResidualAdd.

hi, is there somewhere we can view the feature and constraint of all of TRT supported kernels/plugins? Such as how to insert QDQ or its supported layout?

**Comment by Aktcob** [comment]
> > (and if it's using quantization under the hood, especially for the implicit quantization mode)
> 
> TRT's MHA fusion does not support implicit quantizations yet. Please use explicit quantization instead: Add Q/DQ ops before the two batch gemms in the MHA and also add Q/DQ ops before the ResidualAdd.
> 
> Here is an ugly example: ![2023-07-25 10_19_35-Window](https://private-user-images.githubusercontent.com/53919306/262527659-70a9afaf-69d7-4b14-b8d4-47fe891dbbfa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNDIzMTQsIm5iZiI6MTcxNTE0MjAxNCwicGF0aCI6Ii81MzkxOTMwNi8yNjI1Mjc2NTktNzBhOWFmYWYtNjlkNy00YjE0LWI4ZDQtNDdmZTg5MWRiYmZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA4VDA0MjAxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc5NzM0NThhZDM1Yzk4MTRkNThkYTg0OTNhMTA5OGEwYzJlYmIxMzcwY2FlN2IyMjk5YjExYmZhNTE3ZDUzOGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.o3hm9Q1kYA7VSXelkMSjg606p3WGeGl17-ewRezCpcY)

Hi, I try to convert this onnx to Tensorrt Engine. But there is no kernel name with _mha. 
Tensorrt Version: 8.6.2
![Uploading Screenshot from 2024-05-10 10-53-34.png…]()

**Comment by nvpohanh** [comment]
@Aktcob Could you share your trtexec command and the ONNX? Also, could you try TRT 10.0.1.6 GA release and make sure you have enabled FP16?

**Comment by Aktcob** [comment]
> @Aktcob Could you share your trtexec command and the ONNX? Also, could you try TRT 10.0.1.6 GA release and make sure you have enabled FP16?

@nvpohanh Thanks for reply!
/usr/src/tensorrt/bin/trtexec --onnx=selfattention.onnx --fp16 --iterations=300 --avgRuns=100 --dumpProfile --workspace=1000 --saveEngine=selfattention.engine

I try it on Jetson Orin devkit. So I cannot try it with TRT 10.0.1.6 GA. 

Onnx File: https://wenshu.sankuai.com/file/share/download/35BB3D5239CBFD33D79A5FE4DA4F17BFF7B46221
Password：3xv7u6

**Comment by Aktcob** [comment]
@nvpohanh I try it on another SelfAttention Module without AttentionMask. And there is a kernel with name mha_v2 which fuses matmal + softmax + matmal.
So the question is: how to use mha on selfattention With AttentionMask?

**Comment by ecilay** [comment]
Does Myelin fuses the attention during runtime at engine running or Ahead of time at engine build?

**Comment by Aktcob** [comment]
@nvpohanh Hi, I follow ur example and test my MHA module.

And the onnx model is shown as:
![Image](https://github.com/user-attachments/assets/5ef2b7eb-615e-4873-a8e9-1a51fece9878)

I test it on Jetson Orin Device, Tensorrt version is 8.6.0. And it can be built Tensorrt Engine with mha kernel. I use Nsys to show this.


However, the INT8 mha kernel is much slower than the FP16 mha kernel.(800us vs 200us). How to solve this problem? Thanks for ur reply.

> > (and if it's using quantization under the hood, especially for the implicit quantization mode)
> 
> TRT's MHA fusion does not support implicit quantizations yet. Please use explicit quantization instead: Add Q/DQ ops before the two batch gemms in the MHA and also add Q/DQ ops before the ResidualAdd.
> 
> Here is an ugly example: ![2023-07-25 10_19_35-Window](https://github.com/NVIDIA/TensorRT/assets/53919306/70a9afaf-69d7-4b14-b8d4-47fe891dbbfa)

**Comment by JieShare** [comment]
> [@nvpohanh](https://github.com/nvpohanh) I try it on another SelfAttention Module without AttentionMask. And there is a kernel with name mha_v2 which fuses matmal + softmax + matmal. So the question is: how to use mha on selfattention With AttentionMask?

I encountered the same problem and was wondering if you now know how to better transform self-attention with attention_mask @Aktcob
